File: backend-bars/src/reviews/reviews.controller.ts
import {
  Controller,
  Get,
  Post,
  Put,
  Delete,
  Body,
  Param,
  UseGuards,
  Request,
  HttpCode,
  HttpStatus,
} from '@nestjs/common';
import { JwtAuthGuard } from '../auth/jwt-auth.guard';
import { ReviewsService } from './reviews.service';
import { CreateReviewDto, UpdateReviewDto, OwnerResponseDto } from './dto';

@Controller('reviews')
export class ReviewsController {
  constructor(private readonly reviewsService: ReviewsService) {}

  /**
   * Crear una nueva reseña (requiere autenticación)
   * POST /reviews
   */
  @Post()
  @UseGuards(JwtAuthGuard)
  async create(@Request() req, @Body() createReviewDto: CreateReviewDto) {
    const review = await this.reviewsService.create(req.user.userId, createReviewDto);
    return {
      message: '¡Reseña creada exitosamente!',
      review,
    };
  }

  /**
   * Obtener reseñas de un bar específico (público)
   * GET /reviews/bar/:barId
   */
  @Get('bar/:barId')
  async findByBar(@Param('barId') barId: string) {
    const reviews = await this.reviewsService.findByBar(barId);
    return {
      reviews,
      total: reviews.length,
    };
  }

  /**
   * Obtener estadísticas de reseñas de un bar (público)
   * GET /reviews/bar/:barId/stats
   */
  @Get('bar/:barId/stats')
  async getBarStats(@Param('barId') barId: string) {
    return this.reviewsService.getBarStats(barId);
  }

  /**
   * Obtener mis reseñas (requiere autenticación)
   * GET /reviews/my-reviews
   */
  @Get('my-reviews')
  @UseGuards(JwtAuthGuard)
  async findMyReviews(@Request() req) {
    const reviews = await this.reviewsService.findByUser(req.user.userId);
    return {
      reviews,
      total: reviews.length,
    };
  }

  /**
   * Obtener reseñas de mis bares (para owners)
   * GET /reviews/my-bars
   */
  @Get('my-bars')
  @UseGuards(JwtAuthGuard)
  async findMyBarsReviews(@Request() req) {
    const reviews = await this.reviewsService.findByOwnerBars(req.user.userId);
    return {
      reviews,
      total: reviews.length,
    };
  }

  /**
   * Obtener una reseña por ID (público)
   * GET /reviews/:id
   */
  @Get(':id')
  async findOne(@Param('id') id: string) {
    return this.reviewsService.findOne(id);
  }

  /**
   * Actualizar mi reseña (requiere autenticación)
   * PUT /reviews/:id
   */
  @Put(':id')
  @UseGuards(JwtAuthGuard)
  async update(
    @Param('id') id: string,
    @Request() req,
    @Body() updateReviewDto: UpdateReviewDto,
  ) {
    const review = await this.reviewsService.update(id, req.user.userId, updateReviewDto);
    return {
      message: 'Reseña actualizada exitosamente',
      review,
    };
  }

  /**
   * Eliminar mi reseña (requiere autenticación)
   * DELETE /reviews/:id
   */
  @Delete(':id')
  @UseGuards(JwtAuthGuard)
  @HttpCode(HttpStatus.OK)
  async delete(@Param('id') id: string, @Request() req) {
    await this.reviewsService.delete(id, req.user.userId);
    return {
      message: 'Reseña eliminada exitosamente',
    };
  }

  /**
   * Responder a una reseña (solo owner del bar)
   * POST /reviews/:id/response
   */
  @Post(':id/response')
  @UseGuards(JwtAuthGuard)
  async addOwnerResponse(
    @Param('id') id: string,
    @Request() req,
    @Body() ownerResponseDto: OwnerResponseDto,
  ) {
    const review = await this.reviewsService.addOwnerResponse(
      id,
      req.user.userId,
      ownerResponseDto,
    );
    return {
      message: 'Respuesta añadida exitosamente',
      review,
    };
  }
}

File: front-bars-flutter/lib/modules/reviews/controllers/reviews_controller.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:riverpod_annotation/riverpod_annotation.dart';

import '../models/review_models.dart';
import '../services/reviews_service.dart';

part 'reviews_controller.g.dart';

/// Estado del controlador de reseñas
enum ReviewsStatus { initial, loading, loaded, error }

class ReviewsState {
  final ReviewsStatus status;
  final List<Review> reviews;
  final ReviewStats? stats;
  final String? errorMessage;

  const ReviewsState({
    this.status = ReviewsStatus.initial,
    this.reviews = const [],
    this.stats,
    this.errorMessage,
  });

  ReviewsState copyWith({
    ReviewsStatus? status,
    List<Review>? reviews,
    ReviewStats? stats,
    String? errorMessage,
    bool clearError = false,
  }) {
    return ReviewsState(
      status: status ?? this.status,
      reviews: reviews ?? this.reviews,
      stats: stats ?? this.stats,
      errorMessage: clearError ? null : (errorMessage ?? this.errorMessage),
    );
  }

  bool get isLoading => status == ReviewsStatus.loading;
  bool get hasError => status == ReviewsStatus.error;
  bool get isEmpty => reviews.isEmpty && status == ReviewsStatus.loaded;
}

/// Controlador de reseñas
@riverpod
class ReviewsController extends _$ReviewsController {
  late ReviewsService _reviewsService;

  @override
  ReviewsState build() {
    _reviewsService = ref.watch(reviewsServiceProvider);
    return const ReviewsState();
  }

  /// Cargar reseñas de un bar
  Future<void> loadBarReviews(String barId) async {
    state = state.copyWith(status: ReviewsStatus.loading);

    try {
      final result = await _reviewsService.getBarReviews(barId);

      result.fold(
        (failure) {
          state = state.copyWith(
            status: ReviewsStatus.error,
            errorMessage: failure.message,
          );
        },
        (reviews) {
          state = state.copyWith(
            status: ReviewsStatus.loaded,
            reviews: reviews,
            clearError: true,
          );
        },
      );
    } catch (e) {
      state = state.copyWith(
        status: ReviewsStatus.error,
        errorMessage: 'Error inesperado: ${e.toString()}',
      );
    }
  }

  /// Cargar estadísticas de un bar
  Future<void> loadBarStats(String barId) async {
    try {
      final result = await _reviewsService.getBarStats(barId);

      result.fold(
        (failure) {
          // No cambiar estado de error solo por stats
        },
        (stats) {
          state = state.copyWith(stats: stats);
        },
      );
    } catch (e) {
      // Ignorar errores de stats
    }
  }

  /// Cargar mis reseñas
  Future<void> loadMyReviews() async {
    state = state.copyWith(status: ReviewsStatus.loading);

    try {
      final result = await _reviewsService.getMyReviews();

      result.fold(
        (failure) {
          state = state.copyWith(
            status: ReviewsStatus.error,
            errorMessage: failure.message,
          );
        },
        (reviews) {
          state = state.copyWith(
            status: ReviewsStatus.loaded,
            reviews: reviews,
            clearError: true,
          );
        },
      );
    } catch (e) {
      state = state.copyWith(
        status: ReviewsStatus.error,
        errorMessage: 'Error inesperado: ${e.toString()}',
      );
    }
  }

  /// Cargar reseñas de mis bares (owner)
  Future<void> loadMyBarsReviews() async {
    state = state.copyWith(status: ReviewsStatus.loading);

    try {
      final result = await _reviewsService.getMyBarsReviews();

      result.fold(
        (failure) {
          state = state.copyWith(
            status: ReviewsStatus.error,
            errorMessage: failure.message,
          );
        },
        (reviews) {
          state = state.copyWith(
            status: ReviewsStatus.loaded,
            reviews: reviews,
            clearError: true,
          );
        },
      );
    } catch (e) {
      state = state.copyWith(
        status: ReviewsStatus.error,
        errorMessage: 'Error inesperado: ${e.toString()}',
      );
    }
  }

  /// Crear una nueva reseña
  Future<bool> createReview(CreateReviewDto dto) async {
    state = state.copyWith(status: ReviewsStatus.loading);

    try {
      final result = await _reviewsService.createReview(dto);

      return result.fold(
        (failure) {
          state = state.copyWith(
            status: ReviewsStatus.error,
            errorMessage: failure.message,
          );
          return false;
        },
        (review) {
          state = state.copyWith(
            status: ReviewsStatus.loaded,
            reviews: [review, ...state.reviews],
            clearError: true,
          );
          return true;
        },
      );
    } catch (e) {
      state = state.copyWith(
        status: ReviewsStatus.error,
        errorMessage: 'Error inesperado: ${e.toString()}',
      );
      return false;
    }
  }

  /// Actualizar una reseña
  Future<bool> updateReview(String id, UpdateReviewDto dto) async {
    try {
      final result = await _reviewsService.updateReview(id, dto);

      return result.fold(
        (failure) {
          state = state.copyWith(
            status: ReviewsStatus.error,
            errorMessage: failure.message,
          );
          return false;
        },
        (updatedReview) {
          final updatedReviews = state.reviews.map((r) {
            return r.id == id ? updatedReview : r;
          }).toList();

          state = state.copyWith(
            status: ReviewsStatus.loaded,
            reviews: updatedReviews,
            clearError: true,
          );
          return true;
        },
      );
    } catch (e) {
      state = state.copyWith(
        status: ReviewsStatus.error,
        errorMessage: 'Error inesperado: ${e.toString()}',
      );
      return false;
    }
  }

  /// Eliminar una reseña
  Future<bool> deleteReview(String id) async {
    try {
      final result = await _reviewsService.deleteReview(id);

      return result.fold(
        (failure) {
          state = state.copyWith(
            status: ReviewsStatus.error,
            errorMessage: failure.message,
          );
          return false;
        },
        (_) {
          final updatedReviews = state.reviews.where((r) => r.id != id).toList();

          state = state.copyWith(
            status: ReviewsStatus.loaded,
            reviews: updatedReviews,
            clearError: true,
          );
          return true;
        },
      );
    } catch (e) {
      state = state.copyWith(
        status: ReviewsStatus.error,
        errorMessage: 'Error inesperado: ${e.toString()}',
      );
      return false;
    }
  }

  /// Añadir respuesta del owner
  Future<bool> addOwnerResponse(String id, String response) async {
    try {
      final dto = OwnerResponseDto(response: response);
      final result = await _reviewsService.addOwnerResponse(id, dto);

      return result.fold(
        (failure) {
          state = state.copyWith(
            status: ReviewsStatus.error,
            errorMessage: failure.message,
          );
          return false;
        },
        (updatedReview) {
          final updatedReviews = state.reviews.map((r) {
            return r.id == id ? updatedReview : r;
          }).toList();

          state = state.copyWith(
            status: ReviewsStatus.loaded,
            reviews: updatedReviews,
            clearError: true,
          );
          return true;
        },
      );
    } catch (e) {
      state = state.copyWith(
        status: ReviewsStatus.error,
        errorMessage: 'Error inesperado: ${e.toString()}',
      );
      return false;
    }
  }
}
